crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,764 @@
+        let flakeref = PathRef {
+            path: self.path.clone(),
+            attributes: Default::default(),
+        }
+        .into();
+
+        let attr_path = ["", "floxEnvs", system.as_ref(), "default"]
+            .try_into()
+            .unwrap(); // validated attributes
+
+        FlakeAttribute {
+            flakeref,
+            attr_path,
+            outputs: Default::default(),
         }
     }
 
-    pub fn packages(&self) -> Value {
-        todo!()
+    /// Path to the environment definition file
+    pub fn flox_nix_path(&self) -> PathBuf {
+        self.path.join("pkgs").join("default").join("flox.nix")
     }
 
-    pub async fn install(&self, _packages: &[FloxPackage]) {
-        todo!()
+    /// Build the evironment derivation using nix
+    async fn build(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<(), EnvironmentError2> {
+        debug!("building with nix ....");
+
+        let build = Build {
+            installables: [self.flake_attribute(system).into()].into(),
+            eval: runix::arguments::eval::EvaluationArgs {
+                impure: true.into(),
+                ..Default::default()
+            },
+            build: BuildArgs {
+                // out_link: Some(self.out_link().into()),
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        build
+            .run(nix, &Default::default())
+            .await
+            .map_err(EnvironmentError2::Build)?;
+
+        Ok(())
     }
 
-    pub async fn uninstall(&self, _packages: &[FloxPackage]) {
-        todo!()
+    /// Get a catalog of installed packages from the environment
+    ///
+    /// Evaluated using nix from the environment definition.
+    pub async fn catalog(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<EnvCatalog, EnvironmentError2> {
+        let mut flake_attribute = self.flake_attribute(system);
+        flake_attribute.attr_path.push_attr("catalog").unwrap(); // valid attribute name, should not fail
+
+        let eval = Eval {
+            eval: EvaluationArgs {
+                impure: true.into(),
+                ..Default::default()
+            },
+            eval_args: EvalArgs {
+                installable: Some(flake_attribute.into()),
+                apply: None,
+            },
+            ..Eval::default()
+        };
+
+        let catalog_value: serde_json::Value = eval
+            .run_json(nix, &Default::default())
+            .await
+            .map_err(EnvironmentError2::EvalCatalog)?;
+
+        serde_json::from_value(catalog_value).map_err(EnvironmentError2::ParseCatalog)
     }
 
-    pub async fn upgrade(&self, _packages: &[FloxPackage]) {
+    /// List all Packages in the environment
+    ///
+    /// Currently unused, supposed to be a cleaned up version of [`Environment<_>::catalog`]
+    pub async fn packages(
```

Comment:
Again, when you say "in the environment," do you mean installed into the environment, or the packages provided by the environment?

Reply:
Are those not the same thing?
The environment provides the packages that are installed to it, anything that might come with it by means of `propagatedBuildInputs` and such is not a guaranteed